contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,265 @@
+
+### Proposed Implementation
+
+Since the upgrades using the scripts have been well tested and tried 
+out on several test and production developments, we decided to re-use
+the shell scripts to perform the upgrade tasks. 
+
+The upgrade scripts will be packaged into upgrade container, with an 
+entrypoint script that will invoke the right scripts depending on 
+the object being upgraded. 
+
+A custom resource called `UpgradeTask` will be defined with the 
+details of the object being upgraded, and will in trun be updated
+with the status of the upgrade by the upgrade-container scripts.  
+
+OpenEBS team will release an openebs-upgrade-tools container 
+that contains all the upgrade scripts supported by openebs. 
+
+The workflow followed by administrator is as follows:
+- Define the Custom Resource and a Service account
+- Create a UpgradeTask CR for the object being upgraded 
+- Create a Kubernetes Job with upgrade-container and passing the 
+  UpgradeTask details. 
+- Administrator can query the UpgradeTask to check the status 
+  and result of the upgrade.
+
+### Design Choices/Decisions
+
+This section captures some of the alternative design considered
+and the reasoning behind selecting a certain approach. 
+
+- A generic UpgradeTask CR vs task specific ones like JivaUpgradeTask,
+  CStorVolumeUpgradeTask, etc. Having specific tasks has the following
+  advantages:
+  * Each tasks can have its own spec. the fields may vary depending
+    on the resource being upgraded.
+  * Writing specific operators for each upgrade that operates on a
+    given type of task will make the upgrades more modular. 
+  
+  However, this also means that every time a new resource type is
+  added, another CR needs to be introduced, managed and learned by 
+  the user. This may still be ok. But a similar pattern where 
+  different specs are required is already addressed by the PVC. To 
+  keep the resources management at a minimum, the PVC type, sub-resource
+  spec pattern will be used to specifiy different resources under
+  a generic UpgradeTask CR. 
+
+  Note that, selecting the Generic Task - doesn't preclude from 
+  writing specific upgrade task operators and upgrades task CRs as long
+  as the management of these specific upgrade tasks are completely 
+  management by the operators and user need not interact with them. One
+  possible future implementation could be: The upgrade-operator can 
+  look at the upgrade task (consider it as a claim) and create a 
+  specific upgrade task and bind it. Then the specific upgrade 
+  operator will operate on the their own resources.  
+
+
+### High Level Design
+
+#### UpgradeTask CR Example
+
+Here is an example of UpgradeTask for upgrading Jiva Volume. The 
+conditions are added by the Upgrade Job
+
+```
+apiVersion: openebs.io/v1alpha1
+kind: UpgradeTask
+metadata:
+  name: upgrade-jiva-pv-001
+  namespace: openebs
+spec:
+  fromVersion: 0.9.0
+  toVersion: 1.0.0
+  jivaVolume:
```

Comment:
In case of cstor pools or volumes, would it be something like cstorVolume or cstorPool,
i.e. 
```
cstorVolume:
     pvName: pvc-3d290e5f-7ada-11e9-b8a5-54e1ad4a9dd4
```
In case of cstor pool,
```
cstorPool:
     poolName: cstor-pool-abc-123
```

Reply:
Just curious to know if this schema should also consider control-plane components upgrade ( if upgrading control-plane is only limited to applying the operator.yaml or a patch.yaml ) ?

Reply:
Is it better to have a key like kind/type: jivaVolume or cstorVolume or cstorPool ?

Reply:
Trying to keep the number of resources minimal and make use of the PV pattern to embed different types into a common resource.